File: modules/seamless-recorder/android/src/main/java/expo/modules/seamlessrecorder/SeamlessPlayerModule.kt
package expo.modules.seamlessrecorder

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Queues

class SeamlessPlayerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoSeamlessPlayer")

    AsyncFunction("startBackgroundPlayback") {
      val context = appContext.reactContext ?: return@AsyncFunction null
      PlaybackService.start(context)
      null
    }

    AsyncFunction("stopBackgroundPlayback") {
      val context = appContext.reactContext ?: return@AsyncFunction null
      PlaybackService.stop(context)
      null
    }

    View(SeamlessPlayerView::class) {
      Events("onPlaybackFinished", "onPlaybackError", "onLiveCaughtUp")

      Prop("chunks") { view: SeamlessPlayerView, chunks: List<String> ->
        view.loadChunks(chunks)
      }

      Prop("startPosition") { view: SeamlessPlayerView, positionMs: Long ->
        view.setStartPosition(positionMs)
      }

      Prop("liveMode") { view: SeamlessPlayerView, live: Boolean ->
        view.setLiveMode(live)
      }

      AsyncFunction("play") { view: SeamlessPlayerView ->
        view.play()
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("pause") { view: SeamlessPlayerView ->
        view.pause()
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("getPositionMs") { view: SeamlessPlayerView ->
        view.getPositionMs()
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("getElapsedMs") { view: SeamlessPlayerView ->
        view.getElapsedMs()
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("seekTo") { view: SeamlessPlayerView, positionMs: Long ->
        view.seekTo(positionMs)
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("setSpeed") { view: SeamlessPlayerView, speed: Float ->
        view.setSpeed(speed)
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("appendChunks") { view: SeamlessPlayerView, urls: List<String> ->
        view.appendChunks(urls)
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("getLoadedChunkCount") { view: SeamlessPlayerView ->
        view.getLoadedChunkCount()
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("seekToChunk") { view: SeamlessPlayerView, windowIndex: Int ->
        view.seekToChunk(windowIndex)
      }.runOnQueue(Queues.MAIN)
    }
  }
}

File: modules/seamless-recorder/ios/SeamlessPlayerModule.swift
import ExpoModulesCore

public class SeamlessPlayerModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoSeamlessPlayer")

        // Background playback - on iOS, audio session is configured in the view.
        // These are no-ops to match the Android API surface.
        AsyncFunction("startBackgroundPlayback") { }
        AsyncFunction("stopBackgroundPlayback") { }

        View(SeamlessPlayerView.self) {
            Events("onPlaybackFinished", "onPlaybackError", "onLiveCaughtUp")

            Prop("chunks") { (view: SeamlessPlayerView, chunks: [String]?) in
                if let chunks = chunks {
                    view.loadChunks(chunks)
                }
            }

            Prop("startPosition") { (view: SeamlessPlayerView, positionMs: Int?) in
                view.setStartPosition(Int64(positionMs ?? 0))
            }

            Prop("liveMode") { (view: SeamlessPlayerView, live: Bool?) in
                view.setLiveMode(live ?? false)
            }

            AsyncFunction("play") { (view: SeamlessPlayerView) in
                view.play()
            }.runOnQueue(.main)

            AsyncFunction("pause") { (view: SeamlessPlayerView) in
                view.pause()
            }.runOnQueue(.main)

            AsyncFunction("getPositionMs") { (view: SeamlessPlayerView) -> Int in
                return Int(view.getPositionMs())
            }.runOnQueue(.main)

            AsyncFunction("getElapsedMs") { (view: SeamlessPlayerView) -> Int in
                return Int(view.getElapsedMs())
            }.runOnQueue(.main)

            AsyncFunction("seekTo") { (view: SeamlessPlayerView, positionMs: Int) in
                view.seekTo(Int64(positionMs))
            }.runOnQueue(.main)

            AsyncFunction("setSpeed") { (view: SeamlessPlayerView, speed: Float) in
                view.setSpeed(speed)
            }.runOnQueue(.main)

            AsyncFunction("appendChunks") { (view: SeamlessPlayerView, urls: [String]) in
                view.appendChunks(urls)
            }.runOnQueue(.main)

            AsyncFunction("getLoadedChunkCount") { (view: SeamlessPlayerView) -> Int in
                return view.getLoadedChunkCount()
            }.runOnQueue(.main)

            AsyncFunction("seekToChunk") { (view: SeamlessPlayerView, windowIndex: Int) in
                view.seekToChunk(windowIndex)
            }.runOnQueue(.main)
        }
    }
}
